File: backtester/src/main/scala/signals/SignalsRepository.scala
package ch.xavier
package signals

import cats.effect.IO
import cats.effect.unsafe.implicits.global
import doobie.Transactor
import doobie.implicits.javasql.*
import doobie.implicits.javatime.*
import doobie.syntax.connectionio.toConnectionIOOps
import doobie.syntax.string.toSqlInterpolator
import org.postgresql.util.PSQLException

import java.sql.Timestamp
import java.time.LocalDateTime
import ch.xavier.signals.Signal
import org.slf4j.{Logger, LoggerFactory}

object SignalsRepository {
  private val logger: Logger = LoggerFactory.getLogger("SignalsRepository")

  private val transactor: Transactor[IO] = Transactor.fromDriverManager[IO](
    "org.postgresql.Driver",
    "jdbc:postgresql://localhost:5430/data",
    "root",
    "toor"
  )

  private var cachedSignals: List[Signal] = List()
  
  //Executed the first time in Application
  def getSignals(startTimestamp: Long = 1): List[Signal] =
    if cachedSignals.isEmpty then
      cachedSignals = cacheSignalsOlderThanOneMonth(startTimestamp)
      logger.info(s"${cachedSignals.length} signals are now cached")

    cachedSignals

  private def cacheSignalsOlderThanOneMonth(startTimestamp: Double): List[Signal] =
//    sql"select entry_price, first_target_price, second_target_price, third_target_price, is_long, stop_loss, symbol, timestamp from signals where signal_id = 4" //TEST FOR STRAT
    sql"select entry_price, first_target_price, second_target_price, third_target_price, is_long, stop_loss, symbol, timestamp from signals where signal_id = 297" //BNB 1 Year fake signal
//    sql"select entry_price, first_target_price, second_target_price, third_target_price, is_long, stop_loss, symbol, timestamp from signals where timestamp < ${(System.currentTimeMillis / 1000) - 2419200} and timestamp > $startTimestamp order by timestamp desc"
      .query[Signal]
      .to[List]
      .transact(transactor)
      .unsafeRunSync()
}

File: signals-listener/signals/SignalsRepository.py
import psycopg
from psycopg.rows import dict_row


class SignalsRepository:
    def __init__(self):
        self.db_connection = psycopg.connect(
            'postgresql://root:toor@localhost:5429/data', row_factory=dict_row, autocommit=True)

    def insert_signal(self, signal: dict) -> None:
        with self.db_connection.cursor() as cursor:
            cursor.execute("""INSERT INTO public.signals 
            (entry_price, first_target_price, second_target_price, third_target_price, is_long, stop_loss, symbol, 
            "timestamp") VALUES (%s, %s, %s, %s, %s, %s, %s, %s)""",
                           (signal['entryPrice'], signal['firstTargetPrice'], signal['secondTargetPrice'],
                            signal['thirdTargetPrice'], signal['firstTargetPrice'] > signal['entryPrice'],
                            signal['stopLoss'], signal['symbol'], signal['timestamp']))
